i2c: RTIO driver for stm32

teburd opened the pull request:
Very much a WIP and draft until complete driver for stm32 (targetting v2 to start) using RTIO.

Some ad-hoc commentary here... I'm finding myself copying a lot of logic from i2c_ll_stm32.c and i2c_ll_stm32_v2.c, would it be better to move common code to something like an i2c_ll_stm32_common.c and refactor the existing drivers for this? @erwango your thoughts and opinions would be greatly appreciated on that.

I'm not working towards supporting recovery or i2c target mode just yet but I'm certainly seeing some recurring patterns and would like to perhaps move those into i2c_common.c or i2c_rtio.c to avoid duplicating the work needed to support recovery for sure.

Target mode is really interesting and I'd love to support it, if only for testing, but that will require a bit more thought.

Commit:
i2c: Add a nucleo_h723zg overlay for i2c_ram test

Adds a simple overlay for use with the click shield for nucleo144 board
and the sram 2 click board to test the st i2c driver with the i2c_ram
test.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
i2c: stm32 controller driver using rtio

Adds the simplest possible i2c controller driver for stm32 with the rtio
interface. Some common functionality between the existing driver and the
rtio driver has been moved to a common C file. Currently only interrupt
driven transfers are supported.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

erwango commented:
> Some ad-hoc commentary here... I'm finding myself copying a lot of logic from i2c_ll_stm32.c and i2c_ll_stm32_v2.c, would it be better to move common code to something like an i2c_ll_stm32_common.c and refactor the existing drivers for this? @erwango your thoughts and opinions would be greatly appreciated on that.

I see no issue in refactoring the existing driver to avoid duplication and ease maintenance.

Commit:
rtio: Use CONCAT in place of ## in macros

Found issues when using ## to concatenate tokens in macros where the
resulting symbolic names were including the ## tokens themselves. Using
CONCAT everywhere fixes the issue.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
A first pass review might be worthwhile at this point, I still haven't done a pass on removing code duplication between the drivers here but.... the driver passes my tests now

```
*** Booting Zephyr OS build v3.6.0-940-ge771e9823fa7 ***
Running TESTSUITE i2c_ram
===================================================================
START - test_ram_rtio
submitting write from thread 0x240002b8 addr 7
 PASS - test_ram_rtio in 0.006 seconds
===================================================================
START - test_ram_rtio_isr
timer submitting write, addr e
timer checking write result, submitting read
read command complete
read data complete
 PASS - test_ram_rtio_isr in 0.011 seconds
===================================================================
START - test_ram_transfer
ram using i2c_transfer from thread 0x240002b8 addr 15
 PASS - test_ram_transfer in 0.007 seconds
===================================================================
START - test_ram_write_read
ram using i2c_write and i2c_write_read from thread 0x240002b8 addr 1c
 PASS - test_ram_write_read in 0.008 seconds
===================================================================
TESTSUITE i2c_ram succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [i2c_ram]: pass = 4, fail = 0, skip = 0, total = 4 duration = 0.032 seconds
 - PASS - [i2c_ram.test_ram_rtio] duration = 0.006 seconds
 - PASS - [i2c_ram.test_ram_rtio_isr] duration = 0.011 seconds
 - PASS - [i2c_ram.test_ram_transfer] duration = 0.007 seconds
 - PASS - [i2c_ram.test_ram_write_read] duration = 0.008 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

Commit:
i2c: Add Kconfig HAS_I2C_RTIO

This config should be selected by any driver that supports the RTIO
interface and can then be checked on by things like tests or drivers
that subsequently require the feature exist.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
i2c: sam twihs driver supports rtio

select HAS_I2C_RTIO when sam twihs is selected

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
i2c: RAM test updates to support power management

It's important we test the power management flows when testing i2c

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
i2c: Add a big transfer test to i2c_ram test

The stm32 driver has a branch when the transfer is larger than 256 bytes
and was effectively untested by the i2c ram test in its initial form.
This added large transfer covers the case where an 8 bit byte count
would overflow and require additional state machinery to manage.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
Adding the autoreload support i'm finding some peculiar behavior that i'm diagnosing, but it makes me believe that autoreload doesn't work like the state diagrams suggest in the stm32h723zg reference manual suggests....

![st_i2c_tx](https://github.com/zephyrproject-rtos/zephyr/assets/135839/ab5b5714-5306-4fe2-995b-898a1e321867)

erwango commented on drivers/i2c/i2c_ll_stm32_rtio.c:
```diff
@@ -0,0 +1,520 @@
+
+	i2c_stm32_disable_transfer_interrupts(dev);
+
+	if (LL_I2C_IsEnabledReloadMode(i2c)) {
+		LL_I2C_DisableReloadMode(i2c);
+	}
```
This could be cleaned up if reload is not enabled (or TODOed).

teburd replied:
Yes, I plan on dropping reload support to start but it'll be a little bit before I can get back to this PR. Perhaps early next week I'll try poking around again.

erwango commented on drivers/i2c/i2c_ll_stm32_rtio.c:
```diff
@@ -0,0 +1,520 @@
+		 * so send a stop condition manually
+		 */
+		LL_I2C_GenerateStopCondition(i2c);
+		ret = -EIO;
```
`ret` to be removed

teburd commented:
Dropped for now as autoreload ran into issues on my nucleo h723zg as this was written for whatever reason, likely made a mistake in its implementation or didn't account for something noted in errata/docs somewhere, but moving on for now
